joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/utils/AstGenRunner.scala

```diff
@@ -91,11 +96,128 @@ class AstGenRunner(config: Config, includeFileRegex: String = "") extends AstGen
         val parsedModFile = filterModFile(srcFiles, out)
         val parsed        = filterFiles(srcFiles, out)
         val skipped       = skippedFiles(in, result.toList)
-        GoAstGenRunnerResult(parsedModFile.headOption, parsed, skipped)
+        segregateByModule(config.inputPath, out.toString, parsedModFile, parsed, skipped)
       case Failure(f) =>
         logger.error("\t- running astgen failed!", f)
-        GoAstGenRunnerResult()
+        List()
     }
   }
 
+  /** Segregate all parsed files including go.mod files under separate modules. This will also segregate modules defined
+    * inside another module
+    */
+  private def segregateByModule(
+    inputPath: String,
+    outPath: String,
+    parsedModFiles: List[String],
+    parsedFiles: List[String],
+    skippedFiles: List[String]
+  ): List[GoAstGenRunnerResult] = {
+    val moduleMeta: ModuleMeta =
+      ModuleMeta(inputPath, outPath, None, ListBuffer[String](), ListBuffer[String](), ListBuffer[ModuleMeta]())
+    if (parsedModFiles.size > 0) {
+      parsedModFiles
+        .sortBy(_.split(JFile.separator).length)
+        .foreach(modFile => {
+          moduleMeta.addModFile(modFile, inputPath, outPath)
+        })
+      parsedFiles.foreach(moduleMeta.addParsedFile)
+      skippedFiles.foreach(moduleMeta.addSkippedFile)
+      moduleMeta.getOnlyChilds()
+    } else {
+      parsedFiles.foreach(moduleMeta.addParsedFile)
+      skippedFiles.foreach(moduleMeta.addSkippedFile)
+      moduleMeta.getAllChilds()
+    }
+  }
+
+  private def getParentFolder(path: String): String = {
+    val parent = Paths.get(path).getParent
+    if (parent != null) parent.toString else ""
+  }
+
+  case class ModuleMeta(
+    modulePath: String,
+    outputModulePath: String,
+    modFilePath: Option[String],
+    parsedFiles: ListBuffer[String],
+    skippedFiles: ListBuffer[String],
+    childModules: ListBuffer[ModuleMeta]
+  ) {
+    def addModFile(modFile: String, inputPath: String, outPath: String): Unit = {
+      import scala.util.control.Breaks.*
+      var processed = false
+      breakable {
```

Comment:
Rather use `collectFirst` instead of introducing the break. Collect first can also then return an `Option` which can be used instead of `processed`